### PR TITLE
Add trivia information to SynConst.Measure

### PR DESCRIPTION
**Pull request by dawedawe**
Combined work with @nojaf 

This adds trivia information to SynConst.Measure to help editor tooling dealing with these.

### CODE REVIEW

**Comment by auduchinok** [comment]
Could you please also capture ranges of optional things, like `SynMeasure.One`, so we could distinguish parsed and synthetic measures?

**Comment by dawedawe** [comment]
> Could you please also capture ranges of optional things, like `SynMeasure.One`, so we could distinguish parsed and synthetic measures?

Yes, I'll add a range to SynMeasure.One. As far as I can see, that's the only one missing.

**Comment by nojaf** [comment]
The interesting use-case for `SynMeasure.One` would be something like `let a = 42< 0_1 / m>`.
If we have the range, we can restore the original notation of `1`.

File: src/Compiler/SyntaxTree/SyntaxTree.fsi [hunk 1]
```diff
@@ -193,13 +193,13 @@ type SynMeasure =
     | Seq of measures: SynMeasure list * range: range
 
     /// A division of two units of measure, e.g. 'kg / m'
-    | Divide of measure1: SynMeasure * measure2: SynMeasure * range: range
+    | Divide of measure1: SynMeasure option * measure2: SynMeasure * range: range
 
     /// A power of a unit of measure, e.g. 'kg ^ 2'
     | Power of measure: SynMeasure * power: SynRationalConst * range: range
```
Comment (auduchinok) [inline src/Compiler/SyntaxTree/SyntaxTree.fsi:199]: I guess ranges inside the rational consts are also needed.

File: tests/service/data/SyntaxTree/Measure/Constant - 06.fs.bsl [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+                      ([Power
+                          (Named ([m], (1,3--1,4)), Integer (12345, (1,5--1,10)),
+                           (1,3--1,10))], (1,3--1,10)),
+                    { LessRange = (1,2--1,3)
+                      GreaterRange = (1,10--1,11) }), (1,0--1,11)), (1,0--1,11))],
+          PreXmlDocEmpty, [], None, (1,0--1,11), { LeadingKeyword = None })],
+      (true, true), { ConditionalDirectives = []
+                      CodeComments = [] }, set []))
+
+(1,0)-(1,11) parse warning The declarations in this file will be placed in an implicit module 'Constant - 06' based on the file name 'Constant - 06.fs'. However this is not a valid F# identifier, so the contents will not be accessible from other files. Consider renaming the file or adding a 'module' or 'namespace' declaration at the top of the file.
```
Comment (nojaf) [inline tests/service/data/SyntaxTree/Measure/Constant - 06.fs.bsl:21]: Place the samples inside a module maybe.
This warning isn't relevant to what you are trying to demonstrate.

Comment (dawedawe) [reply tests/service/data/SyntaxTree/Measure/Constant - 06.fs.bsl]: Good idea, thanks.

File: src/Compiler/SyntaxTree/SyntaxTree.fsi (hunk 1 above)
Comment (dawedawe) [reply src/Compiler/SyntaxTree/SyntaxTree.fsi:199]: Good morning @auduchinok. Are you okay with the ranges we now have inside of `SynRationalConst`?
I think, we don't need anything for `SynRationalConst.Negate` as `SynRationalConst.Rational` has the overall range and now also the partial ranges of `numerator`, `denominator`.

Comment (auduchinok) [reply src/Compiler/SyntaxTree/SyntaxTree.fsi:199]: Please capture the range of `Negate` case as well. Using `lhs` for capturing ranges doesn't work well in recovery rules, so if we add them, we'll need to construct the ranges from the ones captures in the children rules.

**Review by T-Gro — APPROVED** [review]

File: src/Compiler/pars.fsy [hunk 3]
```diff
@@ -6003,7 +6005,9 @@ powerType:
 
   | atomTypeOrAnonRecdType INFIX_AT_HAT_OP atomicRationalConstant
      { if $2 <> "^" && $2 <> "^-" then reportParseErrorAt (rhs parseState 2) (FSComp.SR.parsUnexpectedInfixOperator())
-       if $2 = "^-" then SynType.MeasurePower($1, SynRationalConst.Negate($3), lhs parseState)
+       if $2 = "^-" then
+           let m = unionRanges (rhs parseState 2).EndRange (rhs parseState 3)  // include MINUS in Negate range
```
Comment (auduchinok) [inline src/Compiler/pars.fsy:6005]: The same.

File: src/Compiler/pars.fsy [hunk 4]
```diff
@@ -6224,7 +6228,9 @@ measureTypePower:
 
   | measureTypeAtom INFIX_AT_HAT_OP atomicRationalConstant
      { if $2 <> "^" && $2 <> "^-" then reportParseErrorAt (rhs parseState 2) (FSComp.SR.parsUnexpectedOperatorForUnitOfMeasure())
-       if $2 = "^-" then SynMeasure.Power($1, SynRationalConst.Negate($3), lhs parseState)
+       if $2 = "^-" then
+           let m = unionRanges (rhs parseState 2).EndRange (rhs parseState 3)  // include MINUS in Negate range
```
Comment (auduchinok) [inline src/Compiler/pars.fsy:6228]: The same.

File: src/Compiler/pars.fsy [hunk 5]
```diff
@@ -3278,25 +3278,25 @@ rationalConstant:
...
 
   | INT32
     { if snd $1 then errorR(Error(FSComp.SR.lexOutsideThirtyTwoBitSigned(), lhs parseState))
-      SynRationalConst.Integer(fst $1) }
+      SynRationalConst.Integer(fst $1, lhs parseState) }
 
   | MINUS INT32
     { if snd $2 then errorR(Error(FSComp.SR.lexOutsideThirtyTwoBitSigned(), lhs parseState))
-      SynRationalConst.Negate(SynRationalConst.Integer(fst $2)) }
+      SynRationalConst.Negate(SynRationalConst.Integer(fst $2, lhs parseState), lhs parseState) }
```
Comment (auduchinok) [inline src/Compiler/pars.fsy:3295]: This bit looks wrong. `SynRationalConst.Integer` captures the `-` ranges too.

**Review by auduchinok — APPROVED** [review]
Just one comment from my side, otherwise it seems good to go in. Thanks @dawedawe!

File: src/Compiler/pars.fsy (hunk 5 above)
Comment (dawedawe) [reply src/Compiler/pars.fsy]: Thanks, that's right.
It was hidden from my testing as the rule `rationalConstant` is only used inside of `atomicUnsignedRationalConstant`.
But I added a test to be on the safe side.

**Review by psfinaki — APPROVED** [review]

**Pull request merged**